keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -773,6 +775,71 @@ class SqlArtifactRepository(
   }
 
+  override fun getArtifactSummaryInEnvironment(
+    deliveryConfig: DeliveryConfig,
+    environmentName: String,
+    artifactName: String,
+    artifactType: ArtifactType,
+    version: String
+  ): ArtifactSummaryInEnvironment? {
+    return sqlRetry.withRetry(READ) {
+      jooq
+        .select(
+          ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID,
+          ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID,
+          ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION,
+          ENVIRONMENT_ARTIFACT_VERSIONS.DEPLOYED_AT,
+          ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS
+        )
+        .from(ENVIRONMENT_ARTIFACT_VERSIONS)
+        .innerJoin(ENVIRONMENT)
+        .on(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(ENVIRONMENT.UID))
+        .where(ENVIRONMENT.DELIVERY_CONFIG_UID.eq(deliveryConfig.uid))
+        .and(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(
+          select(ENVIRONMENT.UID).from(ENVIRONMENT).where(ENVIRONMENT.NAME.eq(environmentName)))
+        )
+        .and(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID.eq(
+          select(DELIVERY_ARTIFACT.UID)
+            .from(DELIVERY_ARTIFACT)
+            .where(DELIVERY_ARTIFACT.NAME.eq(artifactName))
+            .and(DELIVERY_ARTIFACT.TYPE.eq(artifactType.name)))
+        )
+        .and(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION.eq(version))
+        .orderBy(ENVIRONMENT_ARTIFACT_VERSIONS.DEPLOYED_AT.desc())
+        .limit(1)
+        .fetchOne { (environmentUid, artifactUid, version, deployedAt, promotionStatus) ->
+          val (replacedBy, replacedAt) = when (promotionStatus) {
+            CURRENT.name, PREVIOUS.name -> {
+              jooq
+                .select(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION, ENVIRONMENT_ARTIFACT_VERSIONS.DEPLOYED_AT)
+                .from(ENVIRONMENT_ARTIFACT_VERSIONS)
+                .where(ENVIRONMENT_ARTIFACT_VERSIONS.ENVIRONMENT_UID.eq(environmentUid))
+                .and(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_UID.eq(artifactUid))
+                .and(ENVIRONMENT_ARTIFACT_VERSIONS.ARTIFACT_VERSION.ne(version))
+                .and(ENVIRONMENT_ARTIFACT_VERSIONS.DEPLOYED_AT.isNotNull)
+                .and(ENVIRONMENT_ARTIFACT_VERSIONS.DEPLOYED_AT.greaterThan(deployedAt))
+                .and(ENVIRONMENT_ARTIFACT_VERSIONS.PROMOTION_STATUS.ne(VETOED.name))
+                .orderBy(ENVIRONMENT_ARTIFACT_VERSIONS.DEPLOYED_AT.asc())
+                .limit(1)
+                .fetchOne { (replacedBy, replacedAt) ->
+                  Pair(replacedBy, replacedAt.toInstant(ZoneOffset.UTC))
+                } ?: Pair(null, null)
+            }
+            else -> Pair(null, null)
+          }
+
+          ArtifactSummaryInEnvironment(
+            environment = environmentName,
+            version = version,
+            state = promotionStatus.toLowerCase(),
+            deployedAt = deployedAt.toInstant(ZoneOffset.UTC),
+            replacedAt = replacedAt,
+            replacedBy = replacedBy
```

Comment:
Why do you only do UTC conversion on one of the timestamps?

Reply:
Good question. I'm still learning my way around `LocalDateTime` to/from `Instant` conversions. Should this be using the component's `clock` instead?

Reply:
I think that's the right thing to do...
https://github.com/spinnaker/keel/blob/35e30ffa607c27b3998b722b3ea963cdb93bd458/keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlDeliveryConfigRepository.kt#L518

Reply:
It depends. Do we want the timestamp stored in the database to be Pacific Time? If so we should inject the system default clock and use that. If we want them to be UTC then this would be correct.

Personally I find it quite annoying to have to do time zone math when I'm trying to debug an issue but I'm open to the argument that UTC is less ambiguous.

Reply:
Can we switch from `TIMESTAMP` to `DATETIME` in the database schema? This would encode the timezone in the data in the database.

Reply:
It appears to me that we're storing everything in UTC in the database, despite the fact that we explicitly convert `Instant` to `LocalDateTime` using the current clock.

For example, this code:
https://github.com/spinnaker/keel/blob/d5646dda5821cd927555f911335c2e0ad5f27489/keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlResourceRepository.kt#L284

Results in this record being added to my database a few minutes ago:
```
01E2KCK7DP3HJQJAQDTQX74Y1N	2020-03-09 19:01:04	{"type":"ResourceActuationLaunched","kind":"titus/cluster@v1","id":"titus:cluster:titustestvpc:serverlablpollo","application":"serverlablpollo","plugin":"TitusClusterHandler","tasks":[{"id":"01E30AR0X70N8RZNR4RED8VH05","name":"Upsert server group serverlablpollo in titustestvpc/eu-west-1"},{"id":"01E30AR10QHV54MYTBV3ZG07Z2","name":"Upsert server group serverlablpollo in titustestvpc/us-east-1"},{"id":"01E30AR0V3446ER87CECW56801","name":"Upsert server group serverlablpollo in titustestvpc/us-west-2"}],"timestamp":"2020-03-09T19:01:03.702Z","ignoreRepeatedInHistory":false,"uid":"titus:cluster:titustestvpc:serverlablpollo"}	RESOURCE
```

What am I missing?

Reply:
@robfletcher Sorry, I totally misread your initial question. I'm actually converting both! Lines 825 and 835.

Reply:
What bothers me is that this seems inconsistent with `DeliveryConfigRepository` and `ResourceRepository`. We should standardize how we treat timestamps across the entire codebase

Reply:
I agree, but can we also agree that perhaps that's worth a PR of its own? 😃

Reply:
https://github.com/spinnaker/keel/issues/862